### PR TITLE
4d-minesweeper: init at 2.0

### PR DESCRIPTION
**Pull request by merspieler**
###### Description of changes

4D Minesweeper written in Godot

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@Janik-Haag

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -10651,6 +10651,12 @@
     github = "Mephistophiles";
     githubId = 4850908;
   };
+  merspieler = {
+    email = "merspieler@airmail.cc";
+    name = "merspieler";
+    github = "merspieler";
+    githubId = 29063255;
+  };
```
Comment (Janik-Haag) [inline maintainers/maintainer-list.nix:10659]: this should be it's own commit called: `maintainers: add merspieler`

File: pkgs/games/4d-minesweeper/default.nix [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    export HOME=$TMPDIR
+    # Link the export-templates to the expected location. The --export commands
+    # expects the template-file at .../templates/3.2.3.stable/linux_x11_64_release
+    # with 3.2.3 being the version of godot.
+    mkdir -p $HOME/.local/share/godot
+    ln -s ${godot-export-templates}/share/godot/templates $HOME/.local/share/godot
+    mkdir -p $out/bin/4d-minesweeper
+    cd source/
+    godot-headless --export "Linux/X11" $out/bin/4d-minesweeper/4d-minesweeper
+    runHook postBuild
```
Comment (Janik-Haag) [inline pkgs/games/4d-minesweeper/default.nix:74]:
```suggestion
    runHook preBuild

    # Cannot create file '/homeless-shelter/.config/godot/projects/...'
    export HOME=$TMPDIR
    
    # Link the export-templates to the expected location. The --export commands
    # expects the template-file at .../templates/3.2.3.stable/linux_x11_64_release
    # with 3.2.3 being the version of godot.
    mkdir -p $HOME/.local/share/godot
    ln -s ${godot-export-templates}/share/godot/templates $HOME/.local/share/godot
    
    mkdir -p $out/bin/4d-minesweeper
    cd source/
    godot-headless --export "Linux/X11" $out/bin/4d-minesweeper/4d-minesweeper

    runHook postBuild
```

just some formatting nitpicks

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
Looks good for the most part. But we might want to wait with merging a few more days since #177578 affects this and should redo the godot stuff.

**Comment by Janik-Haag** [comment]
ci fails because of trailing white spaces:
```
pkgs/games/4d-minesweeper/default.nix:
	67: Trailing whitespace
	73: Trailing whitespace
```

**Comment by Janik-Haag** [comment]
`[Update pkgs/games/4d-minesweeper/default.nix](https://github.com/NixOS/nixpkgs/pull/244054/commits/60a4abb4362584909e038fbf7eba2957648187b2)` should be amended locally, so we can keep atomic commits.

**Comment by Janik-Haag** [comment]
nixpkgs-review fails with:
```
error:
       … from call site

         at /home/janik/.cache/nixpkgs-review/pr-244054/nixpkgs/pkgs/top-level/all-packages.nix:36663:21:

        36662|
        36663|   _4d-minesweeper = callPackage ../games/4d-minesweeper { };
             |                     ^
        36664|

       … while calling 'callPackageWith'

         at /home/janik/.cache/nixpkgs-review/pr-244054/nixpkgs/lib/customisation.nix:122:35:

          121|   */
          122|   callPackageWith = autoArgs: fn: args:
             |                                   ^
          123|     let

       error: evaluation aborted with the following error message: 'Function called without required argument "alsaLib" at /home/janik/.cache/nixpkgs-review/pr-244054/nixpkgs/pkgs/games/4d-minesweeper/default.nix:7, did you mean "alsa-lib"?'
https://github.com/NixOS/nixpkgs/pull/244054 failed to build
```
you probably have to change the alsa-lib buildInput

File: pkgs/games/4d-minesweeper/default.nix [hunk 3]
```diff
@@ -0,0 +1,92 @@
+{ lib
+, copyDesktopItems
+, fetchFromGitHub
+, makeDesktopItem
+, stdenv
+, unzip
+, alsaLib
```
Comment (Janik-Haag) [inline pkgs/games/4d-minesweeper/default.nix:7]:
```suggestion
, alsaLib
```

File: pkgs/games/4d-minesweeper/default.nix [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  };
+
+  nativeBuildInputs = [
+    copyDesktopItems
+    godot-headless
+    unzip
+  ];
+
+  buildInputs = [
+    alsaLib
```
Comment (Janik-Haag) [inline pkgs/games/4d-minesweeper/default.nix:45]:
```suggestion
    alsa-lib
```

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
you are currently using an alias, please change that to the actual package

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 244054` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_4d-minesweeper</li>
  </ul>
</details>

**Comment by Janik-Haag** [comment]
when starting the game it throws this error in my terminal:
```
ERROR: Condition "cache_id.size < 1" is true.
   at: _reload_cache (scene/resources/dynamic_font.cpp:908)
```

**Comment by Janik-Haag** [comment]
And the binary is a folder too deep:
```
./result/bin/4d-minesweeper/4d-minesweeper
```
it should be
```
./result/bin/4d-minesweeper
```

File: pkgs/games/4d-minesweeper/default.nix [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    runHook postBuild
+  '';
+
+  dontInstall = true;
+  dontFixup = true;
+  dontStrip = true;
+
+  meta = with lib; {
+    homepage = "https://github.com/gapophustu/4D-Minesweeper";
+    description = "A 4D Minesweeper game written in godot ";
```
Comment (Janik-Haag) [inline pkgs/games/4d-minesweeper/default.nix:87]:
```suggestion
    description = "A 4D Minesweeper game written in godot";
```
just a small nitpick

**Review by Janik-Haag — COMMENTED** [review]
Mostly looks good now, but you should add a a desktop file with `copyDesktopItems` and `makeDesktopItem`.

**Comment by Janik-Haag** [comment]
just grep for `makeDesktopItem` and you will find a lot of examples like `pkgs/applications/science/logic/key/default.nix`

**Comment by superherointj** [comment]
Eval is failing:

> error: evaluation aborted with the following error message: 'Function called without required argument "godot-export-templates" at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-8/pkgs/games/4d-minesweeper/default.nix:10, did you mean "godot3-export-templates"?'

Maybe "godot-export-templates" => "godot3-export-templates"? (I haven't tested.)

**Comment by superherointj** [comment]
> did the naming of godot change?

My point was pointing out eval failed. I have not digged further. The error suggested that. But I cannot say what is the proper solution. Worth checking which Godot version upstream requires. Then feeding that specific version for both godot engine and templates.
